megatron/data/gpt_dataset.py

```diff
@@ -183,8 +188,44 @@ def __getitem__(self, idx):
                 self.doc_idx[doc_index_l],
                 length=offset_l + 1))
             sample = np.concatenate(sample_list)
+        
+        # Code from: https://github.com/EleutherAI/gpt-neox/blob/FIM-clean/megatron/data/gpt2_dataset.py#L109
+        # TODO(Hailey): can merge the code below this line with code above this line.
+        # TODO(Hailey), cont: above already iterates through loop, so just add the permuting in there?
+        sample = np.array(sample, dtype=np.int64)
+        # # print(sample, sample.shape)
+        # # do FIM here, if enabled
+        fim_rate = self.args.fim_rate
+
+        if fim_rate != 0:
+            assert (fim_rate <= 1 and fim_rate >= 0), "FIM rate must be a probability 0 <= rate <= 1"
+
+            eod = self.tokenizer.eod
+
+            segment_breaks = np.argwhere(sample == eod) # split sample by document
+
+            if segment_breaks.shape != (0, 1): # then there is an EOD token in this example
+                curr_start_position = 0
+                for loc in np.nditer(segment_breaks):
+                    # print(loc - curr_start_position, flush=True)
+                    # permute {prefix, suffix, middle} or {suffix, prefix, middle}
+                    # try:
+                    if loc - curr_start_position > 10: # sometimes examples start with EOD or are too short. so avoid this case
+                        sample[curr_start_position:loc], self.np_rng = \
+                            permute(sample[curr_start_position:loc], self.np_rng, self.args, self.tokenizer)
+                    # except ValueError:
+                    #     # print(loc - curr_start_position, flush=True)
+                    #     pass
+
+                    curr_start_position = loc + 1 # jump over the EOD token
+                # TODO: Check that we are not skipping the last subsequence (after the last eod)?
```

Comment:
It seems to me that we might be skipping the last segment, after the last eod token. If there are N eod-tokens that separate the sequence, then these define N+1 segments that should be permuted, IIUC.

Reply:
Ah I think you’re right! 

I had been under the impression that Megatron dataloaders add an EOD to the end of data samples (since they return a seq of len 2049 at this step if the sequence length is set to 2048) but upon a quick check of final tokens in ‘sample’ here it doesn’t seem to be EOD.